scripts/blackjack.coffee

```diff
@@ -0,0 +1,263 @@
+    # objects or if it's just been reconstituted from JSON
+    true
+
+  toString: () ->
+    if @visible
+      "#{@name} of #{@suit}"
+    else
+      "[hidden]"
+
+
+class Hand
+
+  constructor: (cards=[]) ->
+    @cards = classifyCards(cards)
+
+  @fromObject: (hand) -> new Hand(hand.cards)
+
+  getDealt: (card) ->
+    @cards.push(card)
+
+  hardenIfNeeded: () ->
+    if @value() > 21
+      for card in @cards
+        if card.harden()
+          if @value() <= 21
+            return
+
+  isObject: () ->
+    # See Card.isObject, same deal
+    true
+
+  show: () ->
+    for card in @cards
+      card.visible = true
+
+  toString: () ->
+    if @cards.length > 0
+      @cards.toString()
+    else
+      "[Empty hand]"
+
+  value: () ->
+    # old-school loop because .reduce is for some reason turning
+    # everything into strings.
+    total = 0
+    for card in @cards
+      total += card.value
+    return total
+
+class Blackjack
+  constructor: (oldgame = {}) ->
+    @status = oldgame.status or 'playing'
+    @deck = classifyCards(oldgame.deck) or random_deck()
+    @dealer = classifyHand(oldgame.dealer) or new Hand()
+    @player = classifyHand(oldgame.player) or new Hand()
+
+  deal: (to, who="", msg=null, amount=1, hide=false) ->
+    for i in [1..amount]
+      if @deck.length == 0
+        @deck = random_deck()
+        msg?.send "Dealer shuffles a new deck"
+      card = @deck.pop()
+      card.visible = not hide
+      msg?.send "[#{who}] #{card}"
+      to.getDealt(card)
+
+      if to.value() > 21
+        to.hardenIfNeeded()
+      if to.value() > 21
+        @status = 'over'
+        msg?.send "#{who} busts!"
+        return false
+    return true
+
+  initialDeal: (msg=null) ->
+    who = msg?.message?.user?.name or "You"
+    @deal(@dealer, "Dealer", msg)
+    @deal(@dealer, "Dealer", msg, 1, true)
+    @deal(@player, who, msg, 2)
+
+    #  Check for instant winners
+    if @dealer.value() == 21 and @player.value() == 21
+      msg?.send "Blackjacks for EVERYONE!  TIE!"
+      @status = 'over'
+    else if @dealer.value() == 21
+      msg?.send "Dealer blackjack!  You lose!"
+      @status = 'over'
+    else if @player.value() == 21
+      msg?.send "Player blackjack!  You win!"
+      @status = 'over'
+
+  playerStay: (who="", msg=null) ->
+    @dealer.show()
+    msg?.send "Dealer reveals their hand: #{@dealer}"
+    while @dealer.value() < 17
+      @deal(@dealer, "Dealer", msg)
+
+    if @dealer.value() > 21
+      @status = 'over'
+    else if @dealer.value() < @player.value()
+      @status = 'over'
+      msg?.send "Dealer must stay; You win!"
+    else if @dealer.value() == @player.value()
+      @status = 'over'
+      msg?.send "Dealer must stay; tie!"
+    else
+      @status = 'over'
+      msg?.send "Dealer stays; You lose!"
+
+  playing: () ->
+    @status == 'playing'
+
+
+classifyCards = (cards) ->
+  if cards and cards.length > 0 and not cards[0].isObject
```

Comment:
Purposefully not calling it, actually, all I care about is that it's there (because it won't be on non-objects)